File: feeder.rb
require "redis"
require "redis-namespace"
require "json"
require "rss"
require "sinatra"

get %r{/(?<feed>\w+)}, provides: "atom" do |feed|
  redis = Redis::Namespace.new("feeder:#{feed}", redis: Redis.new)

  halt(404) if !redis.exists?("entries")

  atom = RSS::Maker.make("atom") do |maker|
    maker.channel.id = "http://feeder.zynaps.ru/#{feed}"
    maker.channel.author = redis.get("author") || "zynaps@zynaps.ru"
    maker.channel.title = redis.get("title")
    maker.channel.updated = redis.get("updated") || Time.now.to_datetime.rfc3339

    redis.lrange("entries", 0, 49).each do |entry|
      entry = JSON.parse(entry)

      maker.items.new_item do |item|
        item.id = entry["id"]
        item.title = entry["title"]
        if entry["content"]
          item.content.content = entry["content"]
          item.content.type = entry["content_type"] if entry["content_type"]
        end
        item.link = entry["link"]
        item.updated = entry["updated"]
      end
    end
  end

  atom.to_s
end

File: feeder.py
#!/usr/bin/env python3

# TODO: check if we have entry content

import os
import redis
import json
from flask import Flask
from feedgen.feed import FeedGenerator

redis = redis.Redis.from_url(os.getenv('REDIS_URL'))

app = Flask(__name__)

@app.route('/<feed>')
def generate_feed(feed):
    feed_base = "feeder:%s" % feed

    fg = FeedGenerator()

    fg.id("http://feeder.zynaps.ru/%s" % feed)
    fg.title(redis.get('%s:title' % feed_base))
    fg.updated(redis.get('%s:updated' % feed_base).decode())

    for entry in redis.lrange("%s:entries" % feed_base, 0, 49):
        try:
            entry = json.loads(entry.decode())

            fe = fg.add_entry()

            fe.id(entry['id'])
            fe.title(entry['title'])
            fe.content(entry['content'])
            fe.link(href = entry['link'])
            fe.updated(entry['updated'])
        except:
            continue

    return fg.atom_str(pretty = True)

if __name__ == '__main__':
    from waitress import serve
    from paste.translogger import TransLogger
    serve(TransLogger(app), host = '0.0.0.0', port = 4567)
